lib/_http_server.js

```diff
@@ -227,6 +227,13 @@ function writeHead(statusCode, reason, obj) {
 
   var statusLine = 'HTTP/1.1 ' + statusCode + ' ' + this.statusMessage + CRLF;
 
+  // Any response with a 304 (Not Modified) status code is always 
+  // terminated by the first empty line after the header fields, 
+  // regardless of the header fields present in the message, and 
+  // thus cannot contain a message body or 'trailers' octets.
+  if (statusCode === 304 && headers && headers['Trailer'] !== undefined)
```

Comment:
All good insights. #2842 had a mention about throwing. If it is preferred to discard I'll make the change. Also will review _http_outgoing.js Thanks